code/modules/projectiles/guns/energy/special.dm

```diff
@@ -305,15 +305,137 @@
-	ammo_type = list(/obj/item/ammo_casing/energy/toxplasma)
+	ammo_type = list(/obj/item/ammo_casing/energy/toxplasma, /obj/item/ammo_casing/energy/charged_plasma)
 	shaded_charge = 1
 	can_holster = TRUE
+	atom_say_verb = "beeps"
+	bubble_icon = "swarmer"
+	light_color = "#89078E"
+	light_power = 4
+	var/overloaded = FALSE
+	var/warned = FALSE
+	var/charging = FALSE
+	var/mob/living/carbon/holder = null
+
+/obj/item/gun/energy/toxgun/Initialize(mapload)
+	. = ..()
+	START_PROCESSING(SSfastprocess, src)
+
+/obj/item/gun/energy/toxgun/Destroy()
+	STOP_PROCESSING(SSfastprocess, src)
+	UnregisterSignal(src, COMSIG_ITEM_DROPPED)
+	UnregisterSignal(holder, COMSIG_CARBON_SWAP_HANDS)
+	holder = null
+	return ..()
+
+/obj/item/gun/energy/toxgun/process() //keep your finger on the trigger soldier
+	..()
+	if(overloaded)
+		cell.charge -= 0.5 //2.5 per second, 25 every 10 seconds
+		if(cell.charge <= 25 && !warned)
+			warned = TRUE
+			playsound(loc, 'sound/weapons/smg_empty_alarm.ogg', 75, 1)
+			atom_say("Caution, charge low. Forced discharge in under 10 seconds.")
+		if(cell.charge <= 5)
+			discharge()
+
+/obj/item/gun/energy/toxgun/attack_self(mob/living/user)
+	if(overloaded)
+		to_chat(user, "<span class='warning'>[src] is already overloaded!</span>")
+		return
+	if(cell.charge <= 140) //at least 6 seconds of charge time
+		to_chat(user, "<span class='warning'>[src] does not have enough charge to be overloaded.</span>")
+		return
+	to_chat(user, "<span class='notice'>You begin to overload [src].</span>")
+	charging = TRUE
+	if(do_after(user, 25, target = src))
+		select_fire(user)
+		overloaded = TRUE
+		cell.charge -= 125
+		playsound(loc, 'sound/machines/terminal_prompt_confirm.ogg', 75, 1)
+		atom_say("Overloading successful.")
+		set_light(3) //extra visual effect to make it more noticable to user and victims alike
+		RegisterSignal(src, COMSIG_ITEM_DROPPED, /obj/item/gun/energy/toxgun.proc/discharge)
+		holder = user
+		RegisterSignal(holder, COMSIG_CARBON_SWAP_HANDS, /obj/item/gun/energy/toxgun.proc/discharge)
+	charging = FALSE
+
+/obj/item/gun/energy/toxgun/proc/reset_overloaded()
+	select_fire()
+	set_light(0)
+	overloaded = FALSE
+	warned = FALSE
+	UnregisterSignal(src, COMSIG_ITEM_DROPPED)
+	UnregisterSignal(holder, COMSIG_CARBON_SWAP_HANDS)
+
+/obj/item/gun/energy/toxgun/process_fire(atom/target, mob/living/user, message = TRUE, params, zone_override, bonus_spread = 0)
+	if(charging)
+		return
+	return ..()
+
+/obj/item/gun/energy/toxgun/process_chamber()
+	if(overloaded)
+		do_sparks(2, 1, src)
+		reset_overloaded()
+	..()
+	update_icon()
+
+/obj/item/gun/energy/toxgun/emp_act(severity)
+	..()
+	if(prob(100 / severity))
+		if(overloaded)
+			discharge()
+
+/obj/item/gun/energy/toxgun/proc/discharge() //25% of the time, plasma leak. Otherwise, shoot at a random mob / turf nearby. If no proper mob is found when mob is picked, fire at a turf instead
+	reset_overloaded()
+	do_sparks(2, 1, src)
```

Comment:
```suggestion
/obj/item/gun/energy/toxgun/proc/discharge() //25% of the time, plasma leak. Otherwise, shoot at a random mob / turf nearby. If no proper mob is found when mob is picked, fire at a turf instead
	SIGNAL_HANDLER
	reset_overloaded()
	do_sparks(2, 1, src)
```
all procs called by signals should have `SIGNAL_HANDLER` at the start.

Reply:
Done